hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOffloadableOperation.java

```diff
@@ -0,0 +1,298 @@
+
+        final Data finalDataKey = dataKey;
+        final String finalCaller = getCallerUuid();
+        final long finalThreadId = threadId;
+        final long finalCallId = getCallId();
+        final long finalBegin = begin;
+
+        recordStore.localLock(finalDataKey, finalCaller, finalThreadId, finalCallId, Long.MAX_VALUE);
+
+        ops.onStartAsyncOperation(this);
+        getNodeEngine().getExecutionService().execute(executorName, new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    final Map.Entry entry = createMapEntry(dataKey, previousValue);
+                    final Data result = process(entry);
+                    if (!noOp(entry, previousValue)) {
+                        // only case when we need to synchronize back with the partition-thread
+                        Data newValue = toData(entry.getValue());
+
+                        EntryEventType modificationType;
+                        if (entry.getValue() == null) {
+                            modificationType = REMOVED;
+                        } else {
+                            modificationType = (previousValue == null) ? ADDED : UPDATED;
+                        }
+
+                        updateAndUnlock(toData(previousValue), newValue, modificationType, finalCaller, finalThreadId,
+                                result, finalBegin);
+                    } else {
+                        unlockOnly(result, finalCaller, finalThreadId, finalBegin);
+                    }
+                } catch (Throwable t) {
+                    unlockOnly(t, finalCaller, finalThreadId, finalBegin);
+                } finally {
+                    ops.onCompletionAsyncOperation(EntryOffloadableOperation.this);
+                }
+            }
+        });
+    }
+
+    @SuppressWarnings("checkstyle:parameternumber")
+    private void updateAndUnlock(Data previousValue, Data newValue, EntryEventType modificationType, String caller,
+                                 long threadId, final Object result, long now) {
+        EntryOffloadableSetUnlockOperation updateOperation = new EntryOffloadableSetUnlockOperation(name, modificationType,
+                dataKey, previousValue, newValue, caller, threadId, now, entryProcessor.getBackupProcessor());
+
+        updateOperation.setPartitionId(getPartitionId());
+        ops.invokeOnPartition(updateOperation).andThen(new ExecutionCallback<Object>() {
+            @Override
+            public void onResponse(Object response) {
+                getOperationResponseHandler().sendResponse(EntryOffloadableOperation.this, result);
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                // TODO: verify that if t instanceof RetryableException -> the operation root-operation get retried
+                getOperationResponseHandler().sendResponse(EntryOffloadableOperation.this, t);
```

Comment:
I'm aware of that. It holds if the Operation *throws* this exception. I wanted to verify if it is true if the Operation *sends a response* that is a `RetryableException`, as suggested by Peter, since it's this deferred-response case.
I've just verified it and it works fine only because we are doing sendResponse manually.
If an exception was thrown from within the operation, it would not have been retried, since the returnsResponse method returns `false`. Tricky...